### PR TITLE
[WIP] Etcd kubernetes interface

### PR DESCRIPTION
**Pull request by serathius**
Introduces a minimal kubernetes etcd client interface based on [implicit Kubernetes-etcd contract](https://docs.google.com/document/d/1NUZDiJeiIH5vo_FMaTWf0JtrQKCx0kpEaIIuPoj9P6A/edit?usp=sharing).

Goal, improve testability of the contract and prevent further bugs caused by misuse of etcd.

Known and potential issues:
* https://github.com/kubernetes/kubernetes/issues/110210
* Using response revision from non-KV operation (serializable)

/kind feature

```release-note
NONE
```

Proposed interface (in vendor/go.etcd.io/etcd/client/v3/kubernetes.go file) 
```
type Kubernetes interface {
	Get(ctx context.Context, key string, opts GetOptions) (KubernetesGetResponse, error)
	List(ctx context.Context, prefix string, opts ListOptions) (KubernetesListResponse, error)
	Count(ctx context.Context, prefix string) (int64, error)
	OptimisticPut(ctx context.Context, key string, value []byte, opts PutOptions) (KubernetesPutResponse, error)
	OptimisticDelete(ctx context.Context, key string, opts DeleteOptions) (KubernetesDeleteResponse, error)
	Watch(ctx context.Context, key string, opts WatchOptions) (WatchID, WatchChan)
	RequestProgress(ctx context.Context, id WatchID) error
}

type WatchID string

type GetOptions struct {
	Revision int64
}

type ListOptions struct {
	Revision int64
	Limit int64
	Continue string
}

type WatchOptions struct {
	ID WatchID
	Revision int64
	Prefix bool
}

type PutOptions struct {
	ExpectedRevision int64
	LeaseID LeaseID
}

type DeleteOptions struct {
	ExpectedRevision int64
}

type KubernetesGetResponse struct {
	KV *mvccpb.KeyValue
	Revision int64
}

type KubernetesListResponse struct {
	KVs []*mvccpb.KeyValue
	Count int64
	Revision int64
}

type KubernetesPutResponse struct {
	KV *mvccpb.KeyValue
	Succeeded bool
	Revision int64
}

type KubernetesDeleteResponse struct {
	KV *mvccpb.KeyValue
	Succeeded bool
	Revision int64
}
```

Unit tests that validate correctness of etcd storage are passing. Other tests are failing due to dependency failures.

### CODE REVIEW

**Review by logicalhan — COMMENTED** [review]
Ah I like this, this is quite clean.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	Get(ctx context.Context, key string, revision int64) (kv *mvccpb.KeyValue, rev int64, err error)
+	List(ctx context.Context, prefix string, revision, limit int64, fromKey string) (kvs []*mvccpb.KeyValue, count, rev int64, err error)
+	Count(ctx context.Context, prefix string) (int64, error)
+	OptimisticCreate(ctx context.Context, key string, value []byte) (success bool, rev int64, err error)
+	OptimisticUpdate(ctx context.Context, key string, value []byte, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+	OptimisticDelete(ctx context.Context, key string, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+
+	// Consider merging
+	WatchPrefix(ctx context.Context, prefix string, rev int64) WatchChan
+	WatchKey(ctx context.Context, key string, rev int64) WatchChan
```
Comment (wojtek-t) [inline vendor/go.etcd.io/etcd/client/v3/kubernetes.go:24]: What's the purpose of splitting these two?
I would really prefer to have a single `Watch` method with an additional parameter for isRecursive

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	RequestProgress(ctx context.Context) error
+
+	// Deprecated: Using leases breaks the implicit Kubernetes-etcd contract.
```
Comment (wojtek-t) [inline vendor/go.etcd.io/etcd/client/v3/kubernetes.go:27]: Is the method deprecated? Or rather maybe instead we want to change that on the etcd side?
The concept of leases themselves we actually need.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go (hunk 1 above)
Comment (serathius) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go:24]: Why does apiserver establish watch for single object at all? Shouldn't it demultiplex watch calls?

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go (hunk 2 above)
Comment (serathius) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go:27]: Method is deprecated because allows K8s to assign multiple keys to one lease. Leases are revoked in single transaction causing multiple deletion to happen in one revision. Kubernetes needs to either:
* Use one lease per key, changed in https://github.com/kubernetes/kubernetes/issues/47532 due to performance problems.
* Change watch protocol, for example utilize bookmarks as it was proposed in https://github.com/kubernetes/kubernetes/issues/110210#issuecomment-1370779174
* Don't use etcd leases, use K8s leases as it was proposed https://github.com/kubernetes/kubernetes/issues/110210#issuecomment-1405635208

Comment (wojtek-t) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go:26]: I've seen all of those, but those are all k8s-side changes.
Why we're implicitly excluding:
- change etcd to not delete all objects of a single lease in a single transation
?

Comment (serathius) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go:26]: Ok, added.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go (hunk 1 above)
Comment (serathius) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go]: Hmm, makes sense. I agree that it makes sense to disable caching etcd state in apiserver in environments with limited memory, however watch demultiplexing seems like a always net positive thing.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+type Kubernetes interface {
+	Get(ctx context.Context, key string, revision int64) (kv *mvccpb.KeyValue, rev int64, err error)
+	List(ctx context.Context, prefix string, revision, limit int64, fromKey string) (kvs []*mvccpb.KeyValue, count, rev int64, err error)
+	Count(ctx context.Context, prefix string) (int64, error)
+	OptimisticCreate(ctx context.Context, key string, value []byte) (success bool, rev int64, err error)
+	OptimisticUpdate(ctx context.Context, key string, value []byte, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+	OptimisticDelete(ctx context.Context, key string, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+
+	Watch(ctx context.Context, key string, rev int64, prefix bool) WatchChan
+	RequestProgress(ctx context.Context) error
```
Comment (mm4tt) [inline vendor/go.etcd.io/etcd/client/v3/kubernetes.go:23]: Is this per watch progress or per etcd? If the latter, shouldn't it return something (like the current revision)? How does it work?

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	"go.etcd.io/etcd/api/v3/mvccpb"
+)
+
+func NewKubernetes(c *Client) Kubernetes {
+	return &kubernetes{kv: RetryKVClient(c), watcher: c}
+}
+
+type Kubernetes interface {
+	Get(ctx context.Context, key string, revision int64) (kv *mvccpb.KeyValue, rev int64, err error)
+	List(ctx context.Context, prefix string, revision, limit int64, fromKey string) (kvs []*mvccpb.KeyValue, count, rev int64, err error)
```
Comment (mm4tt) [inline vendor/go.etcd.io/etcd/client/v3/kubernetes.go:16]: I'd strongly encourage putting optional arguments into a "ListOptions" struct. Same for other methods.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go (hunk 2 above)
Comment (mm4tt) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go:25]: I'd be careful with calling something deprecated without having a well-agreed-on alternative for that. Especially, for a feature that is currently in use like Leases.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	OptimisticCreate(ctx context.Context, key string, value []byte) (success bool, rev int64, err error)
+	OptimisticUpdate(ctx context.Context, key string, value []byte, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+	OptimisticDelete(ctx context.Context, key string, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+
+	Watch(ctx context.Context, key string, rev int64, prefix bool) WatchChan
+	RequestProgress(ctx context.Context) error
+
+	// Deprecated: Using leases breaks the implicit Kubernetes-etcd contract.
+	// Lease invalidation deletes multiple keys in one transaction breaking Kubernetes watch assumptions about resourceVersion uniqness.
+	OptimisticCreateWithLease(ctx context.Context, key string, value []byte, lease LeaseID) (success bool, rev int64, err error)
```
Comment (mm4tt) [inline vendor/go.etcd.io/etcd/client/v3/kubernetes.go:27]: Will it create a lease if it doesn't exist? Overall, some comments would be useful here and for other methods.

Comment (serathius) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go:27]: I'm considering moving lease manager to etcd. There is no need Kubernetes storage to be aware of leases.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go (hunk 3 above)
Comment (serathius) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go]: > Is this per watch progress or per etcd? 

Yes, but it's per watch stream. Access to watch streams are not exposed to client, they are based on grpc metadata provided on the context. This usually means all watches use the same stream as long they don't override metadata for the grpc. If we are already changing etcd client we might consider exposing this. Added a `WatchID` response.

>  If the latter, shouldn't it return something (like the current revision)? 

Revision is returned to the watcher on the watch channel as it needs to be synchronized to other on watch events.

File: vendor/go.etcd.io/etcd/client/v3/kubernetes.go (hunk 2 above)
Comment (serathius) [reply vendor/go.etcd.io/etcd/client/v3/kubernetes.go]: Done.

**Comment by serathius** [comment]
Updated the interface.

PTAL @logicalhan @mm4tt @wojtek-t

File: go.mod [hunk 6]
```diff
@@ -248,6 +248,7 @@ require (
 )
 
 replace (
+	go.etcd.io/etcd/client/v3 => github.com/serathius/etcd/client/v3 v3.0.0-20230829133936-3a794fbb8970
```
Comment (aojea) [inline go.mod:251]: this is WIP , right?

Comment (serathius) [reply go.mod:251]: Yes, interface needs to be designed on K8s side and then added to etcd client.

Comment (aojea) [reply go.mod:251]: I see, maybe you already know, but if you set WIP on the title the bot will mark  it as not mergable

File: staging/src/k8s.io/apiserver/pkg/storage/continue.go [hunk 7]
```diff
@@ -58,34 +56,18 @@ func DecodeContinue(continueValue, keyPrefix string) (fromKey string, rv int64,
 		if c.ResourceVersion == 0 {
 			return "", 0, ErrInvalidStartRV
 		}
-		if len(c.StartKey) == 0 {
+		if len(c.ContinueKey) == 0 {
 			return "", 0, ErrEmptyStartKey
 		}
-		// defend against path traversal attacks by clients - path.Clean will ensure that startKey cannot
```
Comment (serathius) [inline staging/src/k8s.io/apiserver/pkg/storage/continue.go:64]: Will revert the change

**Comment by dims** [comment]
This work-in-progress PR needs a rebase. please rebase if this PR is still needed for the upcoming release.
